packages/@aws-cdk/aws-synthetics/lib/canary.ts

```diff
@@ -0,0 +1,254 @@
+   * makes up part of the canary ARN, which is included in outbound calls over the internet.
+   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/servicelens_canaries_security.html
+   *
+   * @default A unique name will be generated
+   */
+  readonly canaryName?: string;
+
+  /**
+   * Specify the endpoint that you want the canary code to hit. Alternatively, you can specify
+   * your own canary script to run.
+   *
+   * 🚧 TODO: implement
+   */
+  //readonly test: Test;
+
+}
+
+export class Canary extends cdk.Resource {
+  /**
+   * Execution role associated with this Canary.
+   */
+  public readonly role: iam.IRole;
+
+  /**
+   * Canary ID
+   * @attribute
+   */
+  public readonly canaryId: string;
+
+  /**
+   * Canary State
+   * @attribute
+   */
+  public readonly canaryState: string;
+
+  /**
+   * Canary Name
+   * @attribute
+   */
+  public readonly canaryName: string;
+
+  /**
+   * Bucket where data from each canary run is stored
+   */
+  private readonly artifactBucket: s3.IBucket;
+
+  constructor(scope: cdk.Construct, id: string, props: CanaryProps) {
+    super(scope, id, {
+      physicalName: props.canaryName || cdk.Lazy.stringValue({
+        produce: () => this.generateUniqueName(),
+      }),
+    });
+
+    if (props.canaryName) {
+      this.validateName(props.canaryName);
+    }
+
+    this.artifactBucket = props.artifactBucket ?? new s3.Bucket(this, 'ServiceBucket');
+
+    // Created role will need these policies to run the Canary.
+    // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-synthetics-canary.html#cfn-synthetics-canary-executionrolearn
+    const policy = new iam.PolicyDocument({
+      statements: [
+        new iam.PolicyStatement({
+          resources: ['*'],
+          actions: ['s3:ListAllMyBuckets'],
+        }),
+        new iam.PolicyStatement({
+          resources: [this.artifactBucket.arnForObjects('*')],
+          actions: ['s3:PutObject', 's3:GetBucketLocation'],
+        }),
+        new iam.PolicyStatement({
+          resources: ['*'],
+          actions: ['cloudwatch:PutMetricData'],
+          conditions: {StringEquals: {'cloudwatch:namespace': 'CloudWatchSynthetics'}},
+        }),
+        new iam.PolicyStatement({
+          resources: ['arn:aws:logs:::*'],
+          actions: ['logs:CreateLogStream', 'logs:CreateLogGroup', 'logs:PutLogEvents'],
+        }),
+      ],
+    });
+
+    this.role = props.role ?? new iam.Role(this, 'ServiceRole', {
+      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
+      inlinePolicies: {
+        canaryPolicy: policy,
+      },
+    });
+
+    const resource: CfnCanary = new CfnCanary(this, 'Resource', {
+      artifactS3Location: this.artifactBucket.s3UrlForObject(),
+      executionRoleArn: this.role.roleArn,
+      startCanaryAfterCreation: props.startAfterCreation ?? true,
+      runtimeVersion: 'syn-1.0',
```

Comment:
I am happy to add a property `runtime` that takes in the enum-like class `Runtime`. The only static method in `Runtime` would be `SYN_1_0` for now.

However, I am wondering if the property should be optional here. If the Canary resource comes out with more runtimes, does it make sense to default to `syn-1.0` still? It would be a breaking change to make the property required, so I think it makes sense to discuss before we do this.

I have been told that we should not be future proofing unless there is a clear use case. By adding `runtime` as a property, we are saying that there is a clear use case for additional runtimes in the future. Shouldn't that also mean we make this property required to prepare for additional runtimes? @NetaNir do you have thoughts on this?

Reply:
So if you **do not** have a `runtime` prop now, and you decide to add it later... Then you'll have to introduce it as *optional*. The annoying part of making it optional is that if the "best practice" is to use the latest version, then the default behavior will not conform (not great).

On the other hand, forcing folks to submit a value where there is only one option isn't so great either... But I guess if/when another runtime is added, it'll make sense.

---

Regarding the `SYN_1_0` being the only constant:
- Please, please, please, please do not name it like that. `SYNTHETICS_1_0` is A LOT better.
- There should also be a factory methods that allows folks to wrap an arbitrary string (so they an use a new workflow even if it's not "hardcoded" in the version of CDK Libs they use.

Reply:
`SYN_1_0` is now `SYNTHETICS_1_0`.

I implemented `runtime` as an optional prop after talking to @NetaNir. Reasoning is that we should not be future proofing if it compromises the ergonomics of the current API, and requiring `runtime` would make the user have to specify the only possible property at the moment.

Can continue this conversation if needed in the next CR.